### PR TITLE
fix: only send identify events if traits changed (#46)

### LINKED ISSUE #46: Avoid sending unnecessary 'identify' events
See https://github.com/redhat-developer/vscode-redhat-telemetry/commit/47e9015e20f2b4f374e0239747df545a5997f849

### PR DESCRIPTION
**Pull request by adietish**
fixes #46

### CODE REVIEW

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/service/TelemetryService.java [hunk 1]
```diff
@@ -65,10 +74,14 @@ public void send(TelemetryEvent event) {
     }
 
     private void sendUserInfo() {
-        if (userInfoSent.compareAndSet(false, true)) {
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/service/TelemetryService.java:68]: was: send user info once per IDE session aka instance creation.

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/service/TelemetryService.java [hunk 2]
```diff
@@ -65,10 +74,14 @@ public void send(TelemetryEvent event) {
...
-            doSend(new TelemetryEvent(
+        synchronized(this) {
+            if (!Objects.equals(environment, reportedEnvironment.get())) {
```
[the remark this answers is not in the document]
Comment (adietish) [reply src/main/java/com/redhat/devtools/intellij/telemetry/core/service/TelemetryService.java:78]: Thinking about it, I am convinced that user info that was sent to segment needs to stored on a plugin basis. A change in plugin version needs to be taken into account as env change and sent. I therefore cannot simply ignore the plugin when checking if the env changed. 
Turning this PR to draft and doing this change.

[the remark this answers is not in the document]
Comment (adietish) [reply src/main/java/com/redhat/devtools/intellij/telemetry/core/service/TelemetryService.java:78]: identity in terms of segment is based on specific properties of the environment, namely:
https://github.com/redhat-developer/intellij-redhat-telemetry/blob/main/src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java#L125-L130
```
        putIfNotNull(PROP_LOCALE, environment.getLocale(), properties);
        putIfNotNull(PROP_OS_NAME, environment.getPlatform().getName(), properties);
        putIfNotNull(PROP_OS_DISTRIBUTION, environment.getPlatform().getDistribution(), properties);
        putIfNotNull(PROP_OS_VERSION, environment.getPlatform().getVersion(), properties);
        putIfNotNull(PROP_TIMEZONE, environment.getTimezone(), properties);
```
I'll only check on those to determine if I should send the user info or not.

[the remark this answers is not in the document]
Comment (adietish) [reply src/main/java/com/redhat/devtools/intellij/telemetry/core/service/TelemetryService.java]: Implemented the above in the segment broker like fred did for vscode-redhat-telemetry.

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java [hunk 3]
```diff
@@ -108,26 +110,41 @@ public void send(TelemetryEvent event) {
...
 
     private MessageBuilder toMessage(IdentifyMessage.Builder builder, TelemetryEvent event, Map<String, Object> context) {
+        IdentifyTraits identifyTraits = new IdentifyTraits(
+                environment.getLocale(),
+                environment.getTimezone(),
+                environment.getPlatform().getName(),
+                environment.getPlatform().getVersion(),
+                environment.getPlatform().getDistribution());
+        if (identifyTraitsPersistence != null
+                && identifyTraits.equals(identifyTraitsPersistence.get())) {
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java:132]: now: only send user info (identify traits) if they differ from what we sent last time

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java [hunk 4]
```diff
@@ -108,26 +110,44 @@ public void send(TelemetryEvent event) {
...
 
     private MessageBuilder toMessage(IdentifyMessage.Builder builder, TelemetryEvent event, Map<String, Object> context) {
+        IdentifyTraits identifyTraits = new IdentifyTraits(
+                environment.getLocale(),
+                environment.getTimezone(),
+                environment.getPlatform().getName(),
+                environment.getPlatform().getVersion(),
+                environment.getPlatform().getDistribution());
+        if (identifyTraitsPersistence != null) {
+            if (identifyTraits.equals(identifyTraitsPersistence.get())) {
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java:132]: now: only send user info (identify traits) if they changed compared to the ones we sent last time.

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java [hunk 5]
```diff
@@ -108,26 +110,57 @@ public void send(TelemetryEvent event) {
...
     }
 
     private MessageBuilder toMessage(IdentifyMessage.Builder builder, TelemetryEvent event, Map<String, Object> context) {
+        IdentifyTraits identifyTraits = new IdentifyTraits(
+                environment.getLocale(),
+                environment.getTimezone(),
+                environment.getPlatform().getName(),
+                environment.getPlatform().getVersion(),
+                environment.getPlatform().getDistribution());
+        if (!haveChanged(identifyTraits, identifyTraitsPersistence)) {
```
Comment (adietish) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/service/segment/SegmentBroker.java:131]: now: only send user info (identify traits) if they changed compared to the ones we sent last time.

**Review by jeffmaury — APPROVED** [review]
LGTM

**Pull request merged**